### ISSUE TITLE
Retrieve comments and metadata from page annotations.

### ISSUE THREAD

**Issue by boan-anbo** [ajrcarey/pdfium-render#40]
Hi there! Great lib and I was able to use it to read the PDF! 

I was hoping to work with notes, such as highlight. I can get the `rect` object but couldn't find the Api (perhaps FPDFText_GetBoundedText? I'm not sure as I'm new to Pdfium as well. ) to use the rect of a highlight area to get the highlighted texts.

I was also wondering if it's possible to get the comment text that I type in the pop-up box as well.

Thank you!

**Comment by ajrcarey** [comment]
Hi @boan-anbo, good to hear from you and thanks for opening the issue.

If I remember correctly, Pdfium handles highlighting and comments as annotations. Each annotation has a rectangular bounding area and from the rectangle of a highlight it should be simple enough to read the text on the page that lies within that rectangular region.

I need to look at the Pdfium API to see how text attached to comments is handled.

I'm away at the moment but will look at this for you when I return home over the weekend.

**Comment by ajrcarey** [comment]
PS in the meantime if you have a sample PDF file that includes a highlight and comment you are trying to work with, please attach it to this issue.

**Comment by boan-anbo** [comment]
Thanks so much! And do take your time! I'm attaching a sample pdf (it's written in unicode characters).  
[chn.pdf](https://github.com/ajrcarey/pdfium-render/files/9410554/chn.pdf)

**Comment by ajrcarey** [comment]
Ok, I've had some time to look at this now. Thank you for the sample document, that's excellent. I don't read Chinese, so let me know if there is any content in there that is privileged; otherwise I will include it in the repo as `test/annotations-test.pdf`.

It was already possible to retrieve the text relating to an annotation by looking up the annotation's bounding box, then using this bounding box to retrieve the text using PdfPage:Text::for_rect(), like this:

`page.text()?.for_rect(annotation.bounds()?)`

I have added two new convenience functions to make this simpler, `PdfPageText::for_annotation()` and `PdfPageText::chars_for_annotation()`. The fragment above can now be expressed more easily as:

`page.text()?.for_annotation(&annotation)?`

`PdfPageText::for_annotation()` returns the text linked to the annotation as a string. `PdfPageText::chars_for_annotation()` returns an iterator over the specific glyphs linked to the annotation, if you need more accurate measurement of individual character glyphs.

For the annotations themselves, I have added new functions to let you retrieve the content of the annotation, the annotation's author, and the annotation's creation and modification dates. These are all available via the `PdfPageAnnotationCommon` trait. An example that uses these:

````rust
Pdfium::new(
        Pdfium::bind_to_library(Pdfium::pdfium_platform_library_name_at_path("./"))
            .or_else(|_| Pdfium::bind_to_system_library())?,
    )
    .load_pdf_from_file("test/annotations-test.pdf", None)?
    .pages()
    .iter()
    .enumerate()
    .for_each(|(page_index, page)| {
        // For each page in the document, iterate over the annotations attached to that page.

        println!("=============== Page {} ===============", page_index);

        page.annotations()
            .iter()
            .enumerate()
            .for_each(|(annotation_index, annotation)| {
                println!(
                    "Annotation {} is of type {:?} with bounds {:?}",
                    annotation_index,
                    annotation.annotation_type(),
                    annotation.bounds()
                );

                println!(
                    "{}",
                    page.text().unwrap().for_annotation(&annotation).unwrap()
                );

                println!(
                    "Annotation {} name: {:?}",
                    annotation_index,
                    annotation.name()
                );

                println!(
                    "Annotation {} contents: {:?}",
                    annotation_index,
                    annotation.contents()
                );

                println!(
                    "Annotation {} author: {:?}",
                    annotation_index,
                    annotation.creator()
                );

                println!(
                    "Annotation {} created: {:?}",
                    annotation_index,
                    annotation.creation_date()
                );

                println!(
                    "Annotation {} last modified: {:?}",
                    annotation_index,
                    annotation.modification_date()
                );
            });
    });
````

This example is available in the repo as examples/annotations.rs.

There are more attributes that can be retrieved for each annotation, including the rendered position of the annotation's pop-up and its colors and border types. Let me know if you need these.

Pdfium can create annotations as well as read the properties of existing annotations. Let me know if you need to create annotations as well as read them.

**Comment by ajrcarey** [comment]
Updated `README.md` and `examples/README.md`. Fixed bugs in string key handling in `FPDFAnnot_*` functions in native bindings implementation. Packaged for release as version 0.7.15.

**Comment by ajrcarey** [comment]
Published as 0.7.15. Feel free to reopen the issue if you find any problems with the new functionality, and please open a new issue if you need additional functionality.

**Comment by boan-anbo** [comment]
Thanks so much @ajrcarey ! Haven't found time to go back to this, but will do so. Just wanna express my appreciation for the hard work. More feedback soon.
